File: packages/core/src/kanban.ts
/**
 * The suite kanban (PLAN-KANBAN-UNIFY): one board per workspace plus the
 * suite-wide General board, all stored once in QuantMCP's kanban.db and
 * served by the `plugin:mcp` kanban commands. This module is the typed
 * client every surface renders from — QuantCanvas, the shell drawer and
 * the QuantMCP workspaces page draw the same cards.
 *
 * core.db still carries one entity per card (`core / kanban.card`, id
 * `core:kanban.card:mcp-<card id>`) — a READ mirror for Ctrl+K search and
 * the links table, written by the Rust side on every mutation. Its bus
 * events double as the change signal here. Never write the mirror.
 *
 * In the browser (plain `nuxt dev`) there is no board: `kanbanAvailable()`
 * is false and surfaces say so instead of faking one.
 */

import { on as busOn } from './bus'

/** Sentinel board id: the General board. Not a workspace — no folder, so
 *  its cards cannot be claimed by agents until moved to a workspace. */
export const GENERAL_BOARD_ID = 'general'

export type KanbanColumn = 'plan' | 'work' | 'review' | 'done'

export const KANBAN_COLUMNS: { id: KanbanColumn; label: string }[] = [
  { id: 'plan', label: 'Plan' },
  { id: 'work', label: 'Work' },
  { id: 'review', label: 'Review' },
  { id: 'done', label: 'Done' },
]

export type CardPriority = 'low' | 'medium' | 'high' | 'critical'

export const CARD_PRIORITIES: { id: CardPriority; label: string }[] = [
  { id: 'low', label: 'Low' },
  { id: 'medium', label: 'Medium' },
  { id: 'high', label: 'High' },
  { id: 'critical', label: 'Critical' },
]

/**
 * What the card editor (`QKanbanCardModal`) hands back. The host maps it
 * onto the commands: add for a new card, update + move for an existing one
 * (`update_kanban_card` carries no column).
 */
export interface KanbanCardDraft {
  title: string
  description: string
  column: KanbanColumn
  priority: CardPriority
}

export type CardStatus =
  | 'backlog'
  | 'in_progress'
  | 'awaiting_review'
  | 'approved'
  | 'merged'
  | 'rejected'
  | 'cancelled'

/** A card as the Rust side serialises it (snake_case on purpose). */
export interface KanbanCard {
  id: string
  /** Workspace entity id (`core:workspace:<b36>`) or [`GENERAL_BOARD_ID`]. */
  workspace_id: string
  title: string
  description: string
  column: KanbanColumn
  order: number
  priority: CardPriority
  status: CardStatus
  blocked_by: string[]
  created_at: number
  updated_at: number
  start_approved_at?: number | null
  claimed_at?: number | null
  completed_at?: number | null
  merged_at?: number | null
  agent_id?: string | null
  branch?: string | null
  worktree_path?: string | null
  /** Set once the operator archived the card off the Done column. Such a
   *  card is not in `listKanbanCards` — see `listArchivedKanbanCards`. */
  archived_at?: number | null
  /** Derived by the Rust side from `worktree_path`: the one-liners that
   *  run the app from the card's worktree, one per shell. */
  test_commands?: KanbanTestCommands | null
}

/** `worktree::TestCommands` — the same command for three shells. */
export interface KanbanTestCommands {
  powershell: string
  cmd: string
  bash: string
}

/** The card's core.db mirror entity id — what the links table keys on. */
export function kanbanMirrorId(cardId: string): string {
  return `core:kanban.card:mcp-${cardId}`
}

/**
 * An agent holds this card (claimed or under review): surfaces render it
 * read-only — its worktree and branch live outside the board.
 */
export function cardClaimed(card: KanbanCard): boolean {
  return card.agent_id != null || card.status === 'in_progress' || card.status === 'awaiting_review'
}

export function kanbanAvailable(): boolean {
  return typeof window !== 'undefined' && !!(window as any).__TAURI_INTERNALS__
}

async function tauriInvoke() {
  const { invoke } = await import('@tauri-apps/api/core')
  return invoke
}

export async function listKanbanCards(boardId: string): Promise<KanbanCard[]> {
  if (!kanbanAvailable()) return []
  const invoke = await tauriInvoke()
  return invoke<KanbanCard[]>('plugin:mcp|list_kanban_cards', { workspaceId: boardId })
}

export async function addKanbanCard(input: {
  boardId: string
  title: string
  description?: string
  column?: KanbanColumn
  priority?: CardPriority
  /** Comma-separated card ids. */
  blockedBy?: string
}): Promise<KanbanCard> {
  const invoke = await tauriInvoke()
  return invoke<KanbanCard>('plugin:mcp|add_kanban_card', {
    workspaceId: input.boardId,
    title: input.title,
    description: input.description ?? '',
    column: input.column ?? 'plan',
    priority: input.priority,
    blockedBy: input.blockedBy,
  })
}

export async function updateKanbanCard(
  id: string,
  title: string,
  description: string,
  priority?: CardPriority,
  blockedBy?: string,
): Promise<KanbanCard> {
  const invoke = await tauriInvoke()
  return invoke<KanbanCard>('plugin:mcp|update_kanban_card', {
    id,
    title,
    description,
    priority,
    blockedBy,
  })
}

export async function moveKanbanCard(
  id: string,
  column: KanbanColumn,
  beforeId: string | null = null,
): Promise<void> {
  const invoke = await tauriInvoke()
  await invoke('plugin:mcp|move_kanban_card', { id, column, beforeId })
}

/** Re-home an unclaimed card onto another board (workspace or General). */
export async function moveKanbanCardToBoard(id: string, boardId: string): Promise<KanbanCard> {
  const invoke = await tauriInvoke()
  return invoke<KanbanCard>('plugin:mcp|move_kanban_card_to_workspace', {
    id,
    workspaceId: boardId,
  })
}

export async function deleteKanbanCard(id: string): Promise<void> {
  const invoke = await tauriInvoke()
  await invoke('plugin:mcp|delete_kanban_card', { id })
}

/** Human action from the card dialog; deliberately absent from MCP discovery. */
export async function reviewKanbanCard(id: string, action: 'start' | 'approve' | 'reject', reason?: string): Promise<string> {
  const invoke = await tauriInvoke()
  return invoke<string>('plugin:mcp|review_kanban_card', { id, action, reason })
}

// ── Archive ──
//
// "Archive all" on a board's Done column clears the finished cards off the
// board without deleting them: agents still list them (`list_kanban_cards`
// with archived=true) so earlier work is known and not redone.

/** A board's archived cards, newest archive first. */
export async function listArchivedKanbanCards(boardId: string): Promise<KanbanCard[]> {
  if (!kanbanAvailable()) return []
  const invoke = await tauriInvoke()
  return invoke<KanbanCard[]>('plugin:mcp|list_archived_kanban_cards', { workspaceId: boardId })
}

/** Archive every card in the board's Done column; resolves to the count. */
export async function archiveDoneKanbanCards(boardId: string): Promise<number> {
  const invoke = await tauriInvoke()
  return invoke<number>('plugin:mcp|archive_done_kanban_cards', { workspaceId: boardId })
}

/** Archive one card. Only cards in Done can be archived. */
export async function archiveKanbanCard(id: string): Promise<KanbanCard> {
  const invoke = await tauriInvoke()
  return invoke<KanbanCard>('plugin:mcp|archive_kanban_card', { id })
}

/** Put an archived card back on its board, at the end of its column. */
export async function unarchiveKanbanCard(id: string): Promise<KanbanCard> {
  const invoke = await tauriInvoke()
  return invoke<KanbanCard>('plugin:mcp|unarchive_kanban_card', { id })
}

/** How completed cards leave a board — see the mcp module's settings. On
 *  General nothing merges; the mode only governs the review step. */
export type ApprovalMode = 'auto_apply' | 'approval'

export async function getKanbanApprovalMode(boardId: string): Promise<ApprovalMode> {
  if (!kanbanAvailable()) return 'auto_apply'
  const invoke = await tauriInvoke()
  return invoke<ApprovalMode>('plugin:mcp|get_approval_mode', { workspaceId: boardId })
}

export async function setKanbanApprovalMode(boardId: string, mode: ApprovalMode): Promise<void> {
  const invoke = await tauriInvoke()
  await invoke('plugin:mcp|set_approval_mode', { workspaceId: boardId, mode })
}

/**
 * Fires on any card change in any window — the Rust mirror emits an entity
 * event per mutation; re-list to pick changes up.
 */
export function onKanbanChanged(cb: () => void): () => void {
  if (!kanbanAvailable()) return () => {}
  const offUpsert = busOn<{ kind: string }>('core.entity.upserted', (e) => {
    if (e.payload.kind === 'kanban.card') cb()
  })
  const offDelete = busOn<{ id: string }>('core.entity.deleted', (e) => {
    if (e.payload.id.startsWith('core:kanban.card:')) cb()
  })
  return () => {
    offUpsert()
    offDelete()
  }
}

File: modules/mcp/crate/src/commands/kanban.rs
//! The kanban board commands for the drawer and the module's board page:
//! kanban.db is the one store, one board per workspace plus General.

use crate::kanban_db::{self, KanbanCard, KanbanColumn};
use crate::mcp_server;
use crate::settings;
use crate::AppState;
use std::str::FromStr;

#[tauri::command(async)]
pub(crate) fn list_kanban_cards(state: tauri::State<'_, AppState>, workspace_id: String) -> Result<Vec<KanbanCard>, String> {
    state.kanban_db.list_cards(&workspace_id)
}

// ── Archive ──
//
// The board page's "Archive all" on the Done column and its archive modal.
// Archived cards stay in kanban.db (agents list them with
// `list_kanban_cards archived=true`); they only leave the board.

#[tauri::command(async)]
pub(crate) fn list_archived_kanban_cards(state: tauri::State<'_, AppState>, workspace_id: String) -> Result<Vec<KanbanCard>, String> {
    state.kanban_db.list_archived_cards(&workspace_id)
}

/// Archive one card; the database rejects cards outside Done.
#[tauri::command(async)]
pub(crate) fn archive_kanban_card(state: tauri::State<'_, AppState>, id: String) -> Result<KanbanCard, String> {
    state.kanban_db.archive_card(&id)
}

/// Archive every card in the board's Done column; returns how many moved.
#[tauri::command(async)]
pub(crate) fn archive_done_kanban_cards(state: tauri::State<'_, AppState>, workspace_id: String) -> Result<usize, String> {
    Ok(state.kanban_db.archive_done_cards(&workspace_id)?.len())
}

#[tauri::command(async)]
pub(crate) fn unarchive_kanban_card(state: tauri::State<'_, AppState>, id: String) -> Result<KanbanCard, String> {
    state.kanban_db.unarchive_card(&id)
}

#[tauri::command(async)]
pub(crate) fn add_kanban_card(
    state: tauri::State<'_, AppState>,
    workspace_id: String,
    title: String,
    description: String,
    column: KanbanColumn,
    priority: Option<String>,
    blocked_by: Option<String>,
) -> Result<KanbanCard, String> {
    let pri = priority
        .as_deref()
        .map(kanban_db::CardPriority::from_str)
        .transpose()?
        .unwrap_or(kanban_db::CardPriority::Medium);
    let deps: Vec<String> = blocked_by
        .as_deref()
        .map(|s| s.split(',').map(|x| x.trim().to_string()).filter(|x| !x.is_empty()).collect())
        .unwrap_or_default();
    let card_id = uuid::Uuid::new_v4().to_string();
    state.kanban_db.add_card(&card_id, &workspace_id, &title, &description, &column, &pri, &deps)
}

#[tauri::command(async)]
pub(crate) fn update_kanban_card(
    state: tauri::State<'_, AppState>,
    id: String,
    title: String,
    description: String,
    priority: Option<String>,
    blocked_by: Option<String>,
) -> Result<KanbanCard, String> {
    let pri = priority
        .as_deref()
        .map(kanban_db::CardPriority::from_str)
        .transpose()?;
    let deps: Option<Vec<String>> = blocked_by.as_deref().map(|s| {
        if s.is_empty() { vec![] }
        else { s.split(',').map(|x| x.trim().to_string()).filter(|x| !x.is_empty()).collect() }
    });
    state.kanban_db.update_card(&id, &title, &description, pri.as_ref(), deps.as_deref())
}

#[tauri::command(async)]
pub(crate) fn move_kanban_card(
    app: tauri::AppHandle,
    state: tauri::State<'_, AppState>,
    id: String,
    column: KanbanColumn,
    before_id: Option<String>,
) -> Result<(), String> {
    let card = state.kanban_db.get_card(&id)?.ok_or("Card not found")?;
    if card.workspace_id != kanban_db::GENERAL_BOARD_ID && card.column != column
        && mcp_server::approval_mode_for_card(&app, &card)? == settings::ApprovalMode::Approval
    {
        return Err("Use 'Approve work' or 'Approve result' in the card dialog; dragging cannot skip approval stages.".into());
    }
    state.kanban_db.move_card(&id, &column, before_id.as_deref())
}

/// Human-only actions from the shared card dialog. Not exposed as MCP tools.
#[tauri::command(async)]
pub(crate) fn review_kanban_card(
    app: tauri::AppHandle,
    state: tauri::State<'_, AppState>,
    id: String,
    action: String,
    reason: Option<String>,
) -> Result<String, String> {
    match action.as_str() {
        "start" => {
            state.kanban_db.approve_start(&id)?;
            Ok("Work approved. The agent may now claim this card.".into())
        }
        "approve" => mcp_server::approve_card_from_ui(&app, &state.kanban_db, &id),
        "reject" => {
            let reason = reason.as_deref().unwrap_or("").trim();
            if reason.is_empty() { return Err("Describe the changes you want.".into()); }
            state.kanban_db.reject_card(&id, reason)?;
            Ok("Changes requested. The worktree is preserved.".into())
        }
        _ => Err("Unknown card review action".into()),
    }
}

/// Deleting a claimed card removes its worktree and branch first — the same
/// path as the MCP tool (`mcp_server::delete_card_with_cleanup`). The UI has
/// no line for the cleanup notes, so a warning goes to the log instead.
#[tauri::command(async)]
pub(crate) fn delete_kanban_card(
    app: tauri::AppHandle,
    state: tauri::State<'_, AppState>,
    id: String,
) -> Result<(), String> {
    let note = mcp_server::delete_card_with_cleanup(&app, &state.kanban_db, &id)?;
    if note.contains("WARNING") {
        eprintln!("mcp: delete_kanban_card {id}:{note}");
    }
    Ok(())
}

/// Re-home a card onto another board — `workspace_id` is a workspace entity
/// id or the General sentinel. The drawer's hand-over uses this.
#[tauri::command(async)]
pub(crate) fn move_kanban_card_to_workspace(
    app: tauri::AppHandle,
    state: tauri::State<'_, AppState>,
    id: String,
    workspace_id: String,
) -> Result<KanbanCard, String> {
    if workspace_id != kanban_db::GENERAL_BOARD_ID {
        settings::with_core_db(&app, |conn| {
            qs_core::workspaces::by_id(conn, &workspace_id)
                .ok_or_else(|| format!("Workspace '{}' not registered", workspace_id))
        })?;
    }
    state.kanban_db.move_card_to_board(&id, &workspace_id)
}
